include/albatross/src/models/sparse_gp.hpp

```diff
@@ -137,8 +137,7 @@ class SparseGaussianProcessRegression
     initialize_params();
   };
   SparseGaussianProcessRegression(
-      CovFunc &covariance_function,
-      GrouperFunction &independent_group_function_,
+      CovFunc &covariance_function, GrouperFunction independent_group_function_,
```

Comment:
Can you explain the change of the GrouperFunction from an address to an instance? I don't think its wrong, I just want to understand.

Reply:
Yeah it get's tricky.  Basically the `GrouperFunction` needs to be non-const reference in order to support both passing a function directly as well as passing a function pointer.  It has to do with the way the compiler resolves the templated parameter.  I think it tries to     Here's an example that might illustrate what happens:
```
template <typename Func>
auto get_foo_const_ref(const Func &func) {
    std::cout << __PRETTY_FUNCTION__ << std::endl;
    return func();
}

template <typename Func>
auto get_foo(Func func) {
    std::cout << __PRETTY_FUNCTION__ << std::endl;
    return func();
}

void something() {};

int main() {

    get_foo(something);
    get_foo(&something);
    get_foo([](){something();});    

    get_foo_const_ref(something);
    get_foo_const_ref(&something);
    get_foo_const_ref([](){something();});   

    return 0;
}
```
Which outputs:
```
auto get_foo(Func) [with Func = void (*)()]
auto get_foo(Func) [with Func = void (*)()]
auto get_foo(Func) [with Func = main()::<lambda()>]
auto get_foo_const_ref(const Func&) [with Func = void()]
auto get_foo_const_ref(const Func&) [with Func = void (*)()]
auto get_foo_const_ref(const Func&) [with Func = main()::<lambda()>]
```
Notice that `get_foo` converts `something` into a function pointer where `get_foo_const_ref` does not. @jbangelo might be able to shed light on why it happens.  But in our case we end up storing the function inside `SparseGaussianProcessRegression` and in order to store it we need to make sure it's a pointer so only the non const ref version works.